### ISSUE TITLE
Quantization of pytorch

### ISSUE THREAD

**Issue by garusuke** [ultralytics/ultralytics#7032]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

model= YOLO('yolov8n.pt')
model.export(format='tflite', int8 = True)

This was used for quantization.
However, the resulting inference speed was slower than before quantization.
So I would like to get the result of quantization with pytorch instead of tflite, is there any way?

Also, the quantization of the code above,
1. in which format is the quantization done?
2. is the method dynamic or static?
3. which weights are you quantizing?
4. Is the quantization per-tensor or per-channel (per-axis)?

**Comment by glenn-jocher** [comment]
@garusuke hello! For YOLOv8, `model.export(format='tflite', int8=True)` indeed performs int8 quantization for exporting to the TensorFlow Lite format. Regarding your inquiries:

1. Quantization is applied to the model's weights and activations to convert them from floating-point to 8-bit integers.
2. TensorFlow Lite usually applies post-training quantization, which can be dynamic or static depending on the calibration data provided. By default, it uses dynamic range quantization.
3. All the model's layers that support quantization will have their weights quantized.
4. Quantization in TensorFlow Lite is typically per-tensor quantization.

As for exporting a quantized model in PyTorch format, the current repository primarily supports export to TFLite, ONNX, CoreML, and more, with native PyTorch quantization being a part of PyTorch's ecosystem. You may look into PyTorch's native quantization methods for applying quantization within PyTorch.

Keep in mind that quantization can sometimes slow down inference due to the overhead in converting between types and the need for certain hardware optimizations. For further information on the various modes supported by YOLOv8, please refer to our comprehensive documentation at https://docs.ultralytics.com. 📚

**Comment by garusuke** [comment]
Thanks for the reply.

How can I adapt static quantization?

**Comment by glenn-jocher** [comment]
Hello @garusuke,

To utilize static quantization with YOLOv8, you will need to make use of PyTorch's native quantization tools that support static quantization. It requires calibrating the network with a representative dataset to determine the quantization parameters.

Regarding inference speeds, int8 could be slower than float16 if the hardware is optimized for faster float16 operations, or if the model's implementation leverages specific float16 hardware acceleration, such as Tensor Cores in NVIDIA GPUs. Additionally, the overhead from quantization/dequantization operations might reduce speed gains expected from smaller int8 operations.

For the best performance, it is essential to match the quantization strategy with the capabilities of your hardware. 💡

For more details on YOLOv8's capabilities, you can visit the documentation at https://docs.ultralytics.com.

**Comment by garusuke** [comment]
Thanks for your reply.
I have one last question.

model = YOLO('yolov8n.pt') # load an official model
# export the model
model.export(format='onnx',half=True)

The onnx file of type float16 that you can get with this code is

yolov8_detector = YOLO(model_path)
results = yolov8_detector(frame)
img_annotated = results[0].plot() # overlay bounding box
result_object = results[0] # get coordinates of bounding boxes
bounding_boxes = result_object.boxes.xyxy # get xy coordinates of the box

When executed with

onnxruntime/capi/onnxruntime_inference_collection.py", line 220, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Unexpected input data type. Actual: ( tensor(float)) , expected: (tensor(float16))

error.
Is there a solution?

**Comment by glenn-jocher** [comment]
Hi @garusuke,

It appears that your ONNX runtime expects input data in `float16`, but is receiving `float`. It's necessary to ensure your input data matches the expected data type of your exported model. You can manually cast your input tensors to `float16` before inference to align with the models' data type requirements.

Feel free to browse https://docs.ultralytics.com for more on model handling and exporting. 📘

**Comment by garusuke** [comment]
I Resolved.
Thanks for answering so many questions.

**Comment by glenn-jocher** [comment]
@garusuke you're welcome! If you have any more questions in the future, don't hesitate to ask. Happy coding! 😊👍
